### ISSUE TITLE
JSDoc: no type checking when referencing namespaced function type

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#16238]
**TypeScript Version:**  2.3.4

**Code**

```ts
import React, { StatelessComponent } from 'react';

{
  /**
   * @typedef { { name: string } } FooProps
   * @type { React.StatelessComponent<FooProps> }
   */
  const ReactComponent = props => (
    <div>{props.name}</div>
  );

  <ReactComponent name={1} />; // expected error, name should be string, but got none
}

{
  /**
   * @typedef { { name: string } } FooProps
   * @type { StatelessComponent<FooProps> }
   */
  const ReactComponent = props => (
    <div>{props.name}</div>
  );

  <ReactComponent name={1} />; // error
}

```

**Expected behavior:**
The error should occur when referencing `React.StatelessComponent` in the `@type` annotation.

**Actual behavior:**
The error only occurs when referencing imported `StatelessComponent` in the `@type` annotation.

**Comment by OliverJAsh** [comment]
Another example, not using React:

``` ts
// typings.d.ts
declare namespace Foo {
  interface Bar<T> {
    (param: T): string
  }
}

declare module 'foo' {
  export type Bar<T> = Foo.Bar<T>
}
```

``` ts
// main.js
import { Bar } from 'foo';

{
  /**
   * @typedef { { name: string } } FooProps
   * @type {Bar<FooProps>}
   */
  const foo = x => x;
  foo({ name: 1 }); // error, name should be string
}

{
  /**
   * @typedef { { name: string } } FooProps2
   * @type {Foo.Bar<FooProps2>}
   */
  const foo2 = x => x;
  foo2({ name: 1 }); // expected error, name should be string, but got none
}
```

**Comment by mhegazy** [comment]
This looks fixed on TS 2.5
